state/storage.go

```diff
@@ -988,6 +1223,61 @@ func validateStorageConstraintsAgainstCharm(
 	return nil
 }
 
+func validateCharmStorageCountChange(charmStorage charm.Storage, current, n int) error {
+	action := "attach"
+	absn := n
+	if n < 0 {
+		action = "detach"
+		absn = -absn
+	}
+	gerund := action + "ing"
+	pluralise := ""
+	if absn != 1 {
+		pluralise = "s"
+	}
+
+	count := uint64(current + n)
+	if charmStorage.CountMin == 1 && charmStorage.CountMax == 1 && count != 1 {
+		return errors.Errorf("cannot %s, storage is singular", action)
```

Comment:
yeah, it's a bit clunky. I'll leave it for now, may have another go at it in a follow-up -- otherwise based on feedback